xbmc/filesystem/BlurayDirectory.cpp

```diff
@@ -63,174 +58,479 @@ std::string CBlurayDirectory::GetBlurayID()
     {
-      CLog::Log(LOGDEBUG, "CBlurayDirectory::GetRoot - no menu support, skipping menu entry");
-      return;
+      items.Add(GetTitle(title, title->playlist == static_cast<uint32_t>(mainPlaylist)
+                                    ? g_localizeStrings.Get(25004) /* Main Title */
+                                    : g_localizeStrings.Get(25005) /* Title */));
     }
-
-    path.SetFileName("menu");
-    item = std::make_shared<CFileItem>();
-    item->SetPath(path.Get());
-    item->m_bIsFolder = false;
-    item->SetLabel(g_localizeStrings.Get(25003) /* Menus */);
-    item->SetArt("icon", "DefaultProgram.png");
-    items.Add(item);
+    bd_free_title_info(title);
+  }
 }
 
-bool CBlurayDirectory::GetDirectory(const CURL& url, CFileItemList &items)
+bool CBlurayDirectory::GetDirectory(const CURL& url, CFileItemList& items)
 {
   Dispose();
   m_url = url;
-  std::string root = m_url.GetHostName();
-  std::string file = m_url.GetFileName();
+  std::string root{m_url.GetHostName()};
+  std::string file{m_url.GetFileName()};
   URIUtils::RemoveSlashAtEnd(file);
   URIUtils::RemoveSlashAtEnd(root);
 
   if (!InitializeBluray(root))
     return false;
 
-  if(file == "root")
-    GetRoot(items);
-  else if(file == "root/titles")
-    GetTitles(false, items);
+  // /root                              - get main (length >70% longest) playlists
+  // /root/titles                       - get all playlists
+  // /root/episode/<season>/<episode>   - get playlists that correspond with S<season>E<episode>
+  // /root/episode/all                  - get all episodes
+  if (file == "root")
+  {
+    GetTitles(GetTitlesJob::GET_TITLES_MAIN, items, SortTitlesJob::SORT_TITLES_MOVIE);
+
+    // Add all titles and menu options
+    CDiscDirectoryHelper::AddRootOptions(m_url, items, m_blurayMenuSupport);
+  }
+  else if (file == "root/titles")
+    GetTitles(GetTitlesJob::GET_TITLES_ALL, items, SortTitlesJob::SORT_TITLES_MOVIE);
+  else if (StringUtils::StartsWith(file, "root/episode"))
+  {
+    // Get episodes on disc
+    const std::vector<CVideoInfoTag> episodesOnDisc{CDiscDirectoryHelper::GetEpisodesOnDisc(url)};
+
+    int season{-1};
+    int episode{-1};
+    int episodeIndex{-1};
+    if (file != "root/episode/all")
+    {
+      // Get desired episode from path
+      CRegExp regex{true, CRegExp::autoUtf8, R"((root\/episode\/)(\d+)\/(\d+))"};
+      if (regex.RegFind(file) == -1)
+        return false; // Invalid episode path
+      season = std::stoi(regex.GetMatch(2));
+      episode = std::stoi(regex.GetMatch(3));
+
+      // Check desired episode is on disc
+      const auto& it{
+          std::ranges::find_if(episodesOnDisc, [&season, &episode](const CVideoInfoTag& e)
+                               { return e.m_iSeason == season && e.m_iEpisode == episode; })};
+      if (it == episodesOnDisc.end())
+        return false; // Episode not on disc
+      episodeIndex = std::distance(episodesOnDisc.begin(), it);
```

Comment:
Please add `#include <iterator>` to the header section of this cpp file to guarantee that this function is known in this compile unit.